File: backend/app/api/uploads.py
import re
import uuid
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status
from sqlalchemy.orm import Session
from ..core.config import settings
from ..db import get_db
from ..models import WholesaleUpload
from ..schemas import UploadRead
from ..core.security import require_admin

router = APIRouter(prefix="/api/uploads", tags=["uploads"])

_CONTENT_TYPES = {
    "csv": {"text/csv", "application/csv", "application/vnd.ms-excel"},
    "xlsx": {"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"},
    "pdf": {"application/pdf"},
}


def safe_name(name: str) -> str:
    return re.sub(r"[^A-Za-z0-9._-]", "_", name.rsplit("/", 1)[-1])[:120] or "upload"


def _validate_upload(filename: str, content_type: Optional[str], content: bytes) -> None:
    extension = Path(filename).suffix.lower().lstrip(".")
    allowed = {item.strip().lower() for item in settings.allowed_upload_extensions.split(",")}
    if extension not in allowed or extension not in _CONTENT_TYPES:
        raise HTTPException(status_code=415, detail="This file type is not permitted")
    if content_type not in _CONTENT_TYPES[extension]:
        raise HTTPException(status_code=415, detail="File content type does not match its extension")
    # Do not trust a browser-provided MIME type alone.
    if extension == "pdf" and not content.startswith(b"%PDF-"):
        raise HTTPException(status_code=415, detail="Invalid PDF file")
    if extension == "xlsx" and not content.startswith(b"PK\x03\x04"):
        raise HTTPException(status_code=415, detail="Invalid spreadsheet file")
    if extension == "csv" and b"\x00" in content:
        raise HTTPException(status_code=415, detail="Invalid CSV file")


@router.post("", response_model=list[UploadRead], status_code=status.HTTP_201_CREATED, dependencies=[Depends(require_admin)])
async def upload_wholesale_files(files: list[UploadFile] = File(...), db: Session = Depends(get_db)):
    if not files or len(files) > settings.max_files_per_upload:
        raise HTTPException(status_code=400, detail=f"Upload 1 to {settings.max_files_per_upload} files")
    created = []
    for item in files:
        content = await item.read(settings.max_upload_size_bytes + 1)
        if len(content) > settings.max_upload_size_bytes:
            raise HTTPException(status_code=413, detail=f"{item.filename} exceeds the 25 MB limit")
        filename = safe_name(item.filename or "upload")
        _validate_upload(filename, item.content_type, content)
        storage_key = f"{uuid.uuid4().hex}_{filename}"
        (settings.upload_path / storage_key).write_bytes(content)
        record = WholesaleUpload(original_name=filename, storage_key=storage_key, content_type=item.content_type, size_bytes=len(content))
        db.add(record)
        created.append(record)
    db.commit()
    for record in created:
        db.refresh(record)
    return created

File: backend/src/routes/uploads.js
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const express = require("express");
const multer = require("multer");
const {
  UnsupportedMediaTypeError,
  TooLargeError,
  ValidationError,
} = require("../errors");

const CONTENT_TYPES = {
  csv: ["text/csv", "application/csv", "application/vnd.ms-excel"],
  xlsx: ["application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"],
  pdf: ["application/pdf"],
};

function safeName(name) {
  return (
    String(name || "upload")
      .split("/")
      .pop()
      .replace(/[^A-Za-z0-9._-]/g, "_")
      .slice(0, 120) || "upload"
  );
}

function validateUpload(file, config) {
  const filename = safeName(file.originalname || "upload");
  const extension = path.extname(filename).slice(1).toLowerCase();
  const allowed = new Set(config.allowedUploadExtensions.map((item) => item.toLowerCase()));

  if (!allowed.has(extension) || !Object.prototype.hasOwnProperty.call(CONTENT_TYPES, extension)) {
    throw new UnsupportedMediaTypeError("This file type is not permitted");
  }
  if (!CONTENT_TYPES[extension].includes(file.mimetype)) {
    throw new UnsupportedMediaTypeError("File content type does not match its extension");
  }

  const content = file.buffer;
  if (extension === "pdf" && !content.slice(0, 5).equals(Buffer.from("%PDF-"))) {
    throw new UnsupportedMediaTypeError("Invalid PDF file");
  }
  if (
    extension === "xlsx" &&
    !content.slice(0, 4).equals(Buffer.from([0x50, 0x4b, 0x03, 0x04]))
  ) {
    throw new UnsupportedMediaTypeError("Invalid spreadsheet file");
  }
  if (extension === "csv" && content.includes(0x00)) {
    throw new UnsupportedMediaTypeError("Invalid CSV file");
  }

  return { filename, extension };
}

function runMulter(upload, req, res) {
  return new Promise((resolve, reject) => {
    upload(req, res, (error) => {
      if (!error) {
        resolve();
        return;
      }
      if (error.code === "LIMIT_FILE_SIZE") {
        reject(new TooLargeError("Uploaded file exceeds the 25 MB limit"));
        return;
      }
      reject(error);
    });
  });
}

function createUploadHandler({ config, repository }) {
  const upload = multer({
    storage: multer.memoryStorage(),
    limits: {
      fileSize: config.maxUploadSizeBytes,
      files: config.maxFilesPerUpload,
    },
  }).array("files", config.maxFilesPerUpload);

  return async (req, res) => {
    await runMulter(upload, req, res);
    const files = req.files || [];
    if (!files.length || files.length > config.maxFilesPerUpload) {
      throw new ValidationError(`Upload 1 to ${config.maxFilesPerUpload} files`);
    }

    const created = [];
    for (const file of files) {
      const { filename } = validateUpload(file, config);
      if (file.size > config.maxUploadSizeBytes) {
        throw new TooLargeError(`${filename} exceeds the 25 MB limit`);
      }

      const storageKey = `${crypto.randomUUID().replaceAll("-", "")}_${filename}`;
      fs.writeFileSync(path.join(config.uploadDir, storageKey), file.buffer);
      const record = await repository.createUploadRecord({
        original_name: filename,
        storage_key: storageKey,
        content_type: file.mimetype,
        size_bytes: file.size,
      });

      created.push({
        id: record.id,
        original_name: record.original_name,
        size_bytes: record.size_bytes,
        created_at: record.created_at,
      });
    }

    res.status(201).json(created);
  };
}

function createUploadsRouter({ config, repository }) {
  const router = express.Router();
  const handler = createUploadHandler({ config, repository });
  router.post("/", handler);
  return { router, handler };
}

module.exports = {
  createUploadsRouter,
};
